beebop/utils.py

```diff
@@ -141,45 +138,81 @@ def get_external_clusters_from_file(
     """
     [Finds sample hashes defined by hashes_list in the given external clusters
     file and returns a dictionary of sample hash to external cluster name. If
     there are multiple external clusters listed for a sample, the lowest
-    cluster number is returned]
+    cluster number is returned. If any samples are found but do  not
+    have a cluster assigned, they are returned separately.]
 
     :param previous_query_clustering_file: [filename
     of the project's external clusters file]
     :param hashes_list: [list of sample hashes to find samples for]
     :param external_clusters_prefix: prefix for external cluster name
-    :return dict: [dict of sample hash to lowest numbered-cluster for that
-        sample]
-    """
-    # copy hashes list to keep track of hashes we still need to find samples
-    # for
-    remaining_hashes = hashes_list[:]
-    result = {}
-    with open(previous_query_clustering_file) as f:
-        reader = csv.reader(f, delimiter=",")
-        for row in reader:
-            # We expect two columns in the external clusters csv: the
-            # first contains the sample id (which will be the hash in
-            # the case of our uploaded samples), and the second contains
-            # all the external cluster numbers for the sample, separated
-            # by semicolons
-            sample_id = row[0]
-            if sample_id in remaining_hashes:
-                # Add lowest numeric cluster to dictionary
-                if len(row) > 1:
-                    cluster_no = get_lowest_cluster(row[1])
-                    result[sample_id] = (
-                        f"{external_clusters_prefix}{cluster_no}"
-                    )
-
-                # Remove sample id from remaining hashes to find
-                remaining_hashes.remove(sample_id)
-
-                # Break if no hashes left to find
-                if len(remaining_hashes) == 0:
-                    break
-    return result
+    :return tuple: [dictionary of sample hash to external cluster name,
+        list of sample hashes that were not found]
+    """
+    df, samples_mask = get_df_sample_mask(
+        previous_query_clustering_file, hashes_list
+    )
+    filtered_df = df[samples_mask]
+
+    # Split into found and not found based on NaN values
+    found_mask = filtered_df["Cluster"].notna()
+    not_found_hashes = filtered_df[~found_mask]["sample"].tolist()
+
+    # Process only rows with valid clusters
+    valid_clusters = filtered_df[found_mask]
+    cluster_numbers = valid_clusters["Cluster"].apply(get_lowest_cluster)
+
+    hash_to_cluster_mapping = (
+        external_clusters_prefix + cluster_numbers.astype(str)
+    )
+    hash_to_cluster_mapping.index = valid_clusters["sample"]
+
+    return hash_to_cluster_mapping.to_dict(), not_found_hashes
+
+
+def update_external_clusters_csv(
+    previous_query_clustering_file: str,
+    not_found_q_names: list,
+    external_clusters_not_found: dict,
```

Comment:
I think this is maybe a confusing name for the parameter since those clusters now are found right? And this is the parameter that contains their external cluster values? Maybe change to something like "external_clusters_to_update" or "external_clusters_from_full_db"?

Reply:
yup agree i have changed as requested